docs/proposals/shoot-control-plane-migration/README.md

```diff
@@ -0,0 +1,234 @@
+
+To be more explicit, the following scenarios should be supported.
+
+| Scenario | Source Seed Reachable from Garden Cluster | Source Seed Running | Source Seed Can Reach Anything Outside |
+| ---:| --- | --- | --- |
+| 1 | Yes | Yes/No | Yes/No |
+| 2 | No | Yes | Yes |
+| 3 | No | No | No |
+| 4 | No | Yes | No |
+
+## Non-goal
+
+* The proposed mechanism need not right away implement a solution for migration across regions/availability zones.
+
+## Reuse
+
+* [Machine Controller Manager](https://github.com/gardener/machine-controller-manager) already has the capability to adopt existing machines in the IaaS layer based on proper machine definitions.
+* [Etcd Backup Restore](https://github.com/gardener/etcd-backup-restore) already backs up the shoot etcd which is running in the shoot namespace of the source seed cluster. We can re-use this completely while restoring the shoot etcd on the destination seed cluster.
+
+## Proposed Solution
+
+### Backup and Restoration
+
+The backup and restoration of the following resources that are currently not stored in the shoot apiserver/etcd is not covered by the [reuse](#reuse) mentioned above.
+
+* Resources from the shoot namespace in the seed.
+  * Terraform resources such as configmaps and secrets.
+  * Machine Controller Manager resources such as machinedeployments, machinesets, machines, machineclasses, secrets etc.
+* Resources from the shoot backup namespace in the seed.
+  * Terraform resources such as configmaps and secrets.
+
+All the additional resources listed above will be backed up in the same object store as the what is used for backing up the shoot etcd.
+
+* The backup will be watch-based and not schedule based.
+* The backup will be full backups at longer intervals followed by incremental backups at smaller intervals similar to the etcd continuous/incremental backup.
+* While taking backup it will take care of dependencies. This is needed since while restoring from it, validation checks needs to be passed.
+    
+    * But the granularity of the changes stored at the incremental backups would be at the object level and not the actual field-level changes. This is different from the etcd continuous/incremental backups where the incremental changes are recorded at the field level.
+
+#### Components of the solution
+
+Taking the idea from the framework setup for etcd-backup-restore, we will introduce new component named 
+**shoot-backup-restore** per shoot in seed namespace. This will backup the control plane component of shoot only from seed namespace.
+
+##### shoot-backup-restore implementation
+
+* It will get list of objects to be backed up via configmap shoot-resources.
+
+  * TODO: Come up json format for dependencies configmap
+  * Initial thought (not full proof) is To have nested dependency fields list  
+  * Example configmap content:
+  ```yaml
+  awsmachineclass: spec.sercretRef
```

Comment:
`Machines`, `MachineClasses`, `MachineDeployments`, `MachineSets` should not be copied. The gardener-controller can recreate them and they'll be automatically adopted.

Reply:
As per my knowledge, If we don't copy machines, old machine's provider ID will be lost (@hardikdr @prashanth26 correct me if i'm wrong). On next reconcile it will create spawn new VM and we will have to forget old worker node completely. 
My understanding of requirement is that only control plane is lost, and we might want to talk old  those worker nodes using new control plane .

Reply:
I agreed with @swapnilgm. I think we should be backing all machine objects.

Reply:
MCM already has the features but it needs the right labels and roles which were there in the earlier machine objects. Hence, the need to restore them.

If there is a way to re-create them via shoot-controller with old labels and roles (without having restore machine objects) then I would expect that MCM will re-adopt them. This feature exists.